Source/MLXNN/PositionalEncoding.swift

```diff
@@ -48,67 +37,14 @@ final public class RoPE: Module, UnaryLayer {
         self.scale = scale
     }
 
-    static func cosSinTheta(key: Key) -> (MLXArray, MLXArray) {
-        if let values = cache[key] {
-            return values
-        }
-
-        let D = key.D / 2
-        let positions = MLXArray(key.offset ..< key.N).asType(key.dtype) * key.scale
-        let freqs = exp(-MLXArray(0 ..< D).asType(key.dtype) * (log(key.base) / Float(D)))
-        let theta = positions.reshaped(-1, 1) * freqs.reshaped(1, -1)
-
-        let result = (cos(theta), sin(theta))
-        cache[key] = result
-
-        return result
-    }
-
-    func rope(costheta: MLXArray, sintheta: MLXArray, x: MLXArray) -> MLXArray {
-        let x1 = x[0 ..< (self.dimensions / 2), axis: -1]
-        let x2 = x[(self.dimensions / 2) ..< self.dimensions, axis: -1]
-
-        let rx1 = x1 * costheta - x2 * sintheta
-        let rx2 = x1 * sintheta + x2 * costheta
-
-        let rx: MLXArray
-        if self.dimensions < x.dim(-1) {
-            rx = concatenated([rx1, rx2, x[self.dimensions..., axis: -1]], axis: -1)
-        } else {
-            rx = concatenated([rx1, rx2], axis: -1)
-        }
-        return rx
-    }
-
-    func traditionalRope(costheta: MLXArray, sintheta: MLXArray, x: MLXArray) -> MLXArray {
-        let x1 = x[stride: 2, axis: -1]
-        let x2 = x[from: 1, stride: 2, axis: -1]
-
-        let rx1 = x1 * costheta - x2 * sintheta
-        let rx2 = x1 * sintheta + x2 * costheta
-
-        if dimensions < x.dim(-1) {
-            fatalError("RoPE doesn't implement partial traditional application")
-        }
-
-        let rx = concatenated(
-            [expandedDimensions(rx1, axis: -1), expandedDimensions(rx2, axis: -1)], axis: -1)
-
-        return rx
-    }
```

Comment:
this is all replaced with `mlx_fast_rope`

Reply:
Beautiful!